services/src/main/java/org/keycloak/services/util/DPoPUtil.java

```diff
@@ -0,0 +1,401 @@
+            throw new VerificationException("DPoP header verification failure");
+        }
+
+        if (!DPOP_JWT_HEADER_TYPE.equals(header.getType())) {
+            throw new VerificationException("Invalid or missing type in DPoP header: " + header.getType());
+        }
+
+        String algorithm = header.getAlgorithm().name();
+
+        if (!DPOP_SUPPORTED_ALGS.contains(algorithm)) {
+            throw new VerificationException("Unsupported DPoP algorithm: " + header.getAlgorithm());
+        }
+
+        JWK jwk = header.getKey();
+        KeyWrapper key;
+
+        if (jwk == null) {
+            throw new VerificationException("No JWK in DPoP header");
+        } else {
+            key = JWKSUtils.getKeyWrapper(jwk);
+            if (key.getPublicKey() == null) {
+                throw new VerificationException("No public key in DPoP header");
+            }
+            if (key.getPrivateKey() != null) {
+                throw new VerificationException("Private key is present in DPoP header");
+            }
+        }
+
+        key.setAlgorithm(header.getAlgorithm().name());
+
+        SignatureVerifierContext signatureVerifier = session.getProvider(SignatureProvider.class, algorithm).verifier(key);
+        verifier.verifierContext(signatureVerifier);
+        verifier.withChecks(
+                DPoPClaimsCheck.INSTANCE,
+                new DPoPHTTPCheck(uri, method),
+                new DPoPIsActiveCheck(session, lifetime, clockSkew),
+                new DPoPReplayCheck(session, lifetime));
+
+        if (accessToken != null) {
+            verifier.withChecks(new DPoPAccessTokenHashCheck(accessToken));
+        }
+
+        try {
+            DPoP dPoP = verifier.verify().getToken();
+            dPoP.setThumbprint(JWKSUtils.computeThumbprint(jwk));
+            return dPoP;
+        } catch (DPoPVerificationException ex) {
+            throw ex;
+        } catch (VerificationException ex) {
+            throw new VerificationException("DPoP verification failure", ex);
+        }
+
+    }
+
+    public static void validateBinding(AccessToken token, DPoP dPoP) throws VerificationException {
+        try {
+            TokenVerifier.createWithoutSignature(token)
+                    .withChecks(new DPoPUtil.DPoPBindingCheck(dPoP))
+                    .verify();
+        } catch (TokenVerificationException ex) {
+            throw ex;
+        } catch (VerificationException ex) {
+            throw new VerificationException("Token verification failure", ex);
+        }
+    }
+
+    public static void bindToken(AccessToken token, String thumbprint) {
+        AccessToken.Confirmation confirmation = token.getConfirmation();
+
+        if (confirmation == null) {
+            confirmation = new AccessToken.Confirmation();
+            token.setConfirmation(confirmation);
+        }
+
+        confirmation.setKeyThumbprint(thumbprint);
+    }
+
+    public static void bindToken(AccessToken token, DPoP dPoP) {
```

Comment:
Can we remove this method and call `public static void bindToken(AccessToken token, String thumbprint)`  directly?